AMD GPUs produce incorrect results

For the exact same setup, a Pop III simulation on Setonix GPUs show a very different evolution as compared to CPUs. The initial density projections on both are identical:

![plt00000_Projection_z_gasDensity_gasDensity](https://github.com/quokka-astro/quokka/assets/34922596/7f65a7c7-ea83-4a0f-acd7-4eb7c995d4f4)

However, after a few 100 steps, GPU projections look weird whereas CPU projections look more reasonable. GPU projection:

![plt00900_Projection_z_gasDensity_gasDensity](https://github.com/quokka-astro/quokka/assets/34922596/5ce599dc-c4f9-4d86-93d7-a9b50a72fc5c)

CPU projection:

![plt00900_Projection_z_gasDensity_gasDensity](https://github.com/quokka-astro/quokka/assets/34922596/85530bac-4560-4edb-9e6e-de1c8ea23970)

GPU run is much faster, but it needs flux correction in several cells in each timestep almost from t = 0, whereas CPU run doesn't need any flux corrections. At some point, the GPU run aborts because of rho = 0 or NAN during regridding.

There seems to be some issue with Setonix GPUs.

> `fcompare` might be a better tool to do a direct comparison between two plotfiles: https://github.com/AMReX-Codes/amrex/blob/development/Tools/Plotfile/fcompare.cpp
> 
> You can compile it by just typing `make` inside `extern/amrex/Tools/Plotfile`.

```
fgradient.cpp:9:10: fatal error: filesystem: No such file or directory
 #include <filesystem>
          ^~~~~~~~~~~~
compilation terminated.
make: *** [../../Tools/GNUMake/Make.rules:89: fgradient.gnu.x86-milan.ex] Error 1
```

> > `fcompare` might be a better tool to do a direct comparison between two plotfiles: https://github.com/AMReX-Codes/amrex/blob/development/Tools/Plotfile/fcompare.cpp
> > You can compile it by just typing `make` inside `extern/amrex/Tools/Plotfile`.
> 
> ```
> fgradient.cpp:9:10: fatal error: filesystem: No such file or directory
>  #include <filesystem>
>           ^~~~~~~~~~~~
> compilation terminated.
> make: *** [../../Tools/GNUMake/Make.rules:89: fgradient.gnu.x86-milan.ex] Error 1
> ```

Can you replace that line with:
```
#if __has_include(<filesystem>)
#include <filesystem>
#elif __has_include(<experimental/filesystem>)
#include <experimental/filesystem>
namespace std
{
namespace filesystem = experimental::filesystem;
}
#endif
```

The memory errors are fixed by adding`-mllvm -amdgpu-function-calls=true` to the compiler flags.

This works around AMDGPU compiler bugs when building very large kernels (the primordial chemistry network, the larger nuclear networks).

There is still a real bug (at least, it also appears on CPU runs) where, at timestep 58, FOFC happens and then then VODE fails to integrate the network (from @psharda):

```
Coarse STEP 58 at t = 2277615624 (2.277615624e-05%) starts ...
	>> Using global timestep on this coarse step (estimated work ratio: 1).
[Level 0 step 58] ADVANCE with time = 2277615624 dt = 228761562.4
[FOFC-1] flux correcting 1 cells on level 0
Coordinates: (-1.215046875e+18, -2.719390625e+18, 8.67890625e+17):  At cell (21,8,39) in Box ((0,0,32) (31,31,63) (0,0,0)): -7.6231090543352448e-24, -1.069268058649304e-17, 7.410320892898785e-18, 3.7609666350373797e-18, 5.3526763892812115e-10, 5.2902546443125309e-10, -2.8030653010662632e-33, -5.1467865827886286e-30, -5.8149295047562081e-24, -3.1857928751784568e-37, -2.5538870986621986e-34, -3.4809871986548917e-28, -6.7835317008872414e-39, -2.5744653822924568e-41, -4.5308774871340948e-27, -6.4171286356493599e-43, -1.3888196456542485e-30, -9.2237860360589853e-74, -5.6154994834351863e-67, -1.8032940347070311e-24
[FOFC-2] flux correcting 1 cells on level 0
Coordinates: (-1.215046875e+18, -2.719390625e+18, 8.67890625e+17):  At cell (21,8,39) in Box ((0,0,32) (31,31,63) (0,0,0)): -4.4439055516774191e-24, -9.7063852770399433e-18, 9.2891283062859141e-18, 3.8792162629719635e-18, 5.3612761960843476e-10, 5.2929401031211604e-10, -1.6338050081804726e-33, -2.9998038185290005e-30, -3.3898238659059865e-24, -1.8648566265595356e-37, -1.4894234044896535e-34, -2.0292483481208408e-28, -3.98069827049991e-39, -1.5061288633265628e-41, -2.6413500902039217e-27, -3.7645391218040064e-43, -8.0955352015117408e-31, -5.3722206263769706e-74, -3.102252662899324e-67, -1.0512335997061365e-24
DVODE: corrector convergence failed repeatedly or with abs(H) = HMIN
[ERROR] integration failed in net
istate = -6
zone = (0, 0, 0)
time = 0
dt = 114380781.1951238
temp start = 84.62999999999991
xn start = 0.007969664777026481 0.007969467582489294 8999.891176964589 4.929760939497877e-10 1.977453278654315e-07 0.269463080239167 5.254014144099686e-12 1.992932720226845e-14 3.506310230602529 3.313755276663732e-16 0.0007166261480667435 3.569520310420554e-47 2.119713584959603e-40 697.9645364035052
dens current = 1.974508903110628e-20
temp current = -nan
xn current = 0.007969664777026481 0.007969467582489294 8999.891176964589 4.929760939497877e-10 1.977453278654315e-07 0.269463080239167 5.254014144099686e-12 1.992932720226845e-14 3.506310230602529 3.313755276663732e-16 0.0007166261480667435 3.569520310420554e-47 2.119713584959603e-40 697.9645364035052
energy generated = 8676380294.558048
amrex::Abort::4::VODE integration was unsuccessful! !!!
```

--- a/.github/workflows/dependencies/dependencies_hip.sh
+++ b/.github/workflows/dependencies/dependencies_hip.sh
@@ -22,7 +22,7 @@ sudo mkdir --parents --mode=0755 /etc/apt/keyrings
 wget https://repo.radeon.com/rocm/rocm.gpg.key -O - | \
     gpg --dearmor | sudo tee /etc/apt/keyrings/rocm.gpg > /dev/null
 
-for ver in 5.7.1 6.3; do
+for ver in 5.7.1 6.3.1; do
 echo "deb [arch=amd64 signed-by=/etc/apt/keyrings/rocm.gpg] https://repo.radeon.com/rocm/apt/$ver focal main" \
     | sudo tee --append /etc/apt/sources.list.d/rocm.list
 done
@@ -43,12 +43,12 @@ sudo apt-get install -y --no-install-recommends \
     libnuma-dev     \
     libopenmpi-dev  \
     openmpi-bin     \
-    rocm-dev5.7.1        \
-    roctracer-dev5.7.1   \
-    rocprofiler-dev5.7.1 \
-    rocrand-dev5.7.1     \
-    rocprim-dev5.7.1     \
-    rocsparse-dev5.7.1
+    rocm-dev6.3.1        \
+    roctracer-dev6.3.1   \
+    rocprofiler-dev6.3.1 \
+    rocrand-dev6.3.1     \
+    rocprim-dev6.3.1     \
+    rocsparse-dev6.3.1
 
 # activate
 #

--- a/.github/workflows/dependencies/dependencies_hip.sh
+++ b/.github/workflows/dependencies/dependencies_hip.sh
@@ -47,6 +47,7 @@ sudo apt-get install -y --no-install-recommends \
     roctracer-dev6.3.1   \
     rocprofiler-dev6.3.1 \
     rocrand-dev6.3.1     \
+    hiprand-dev6.3.1     \
     rocprim-dev6.3.1     \
     rocsparse-dev6.3.1
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -64,7 +64,7 @@ if(AMReX_GPU_BACKEND MATCHES "HIP")
   string(REPLACE "HIP version: " "" hipcc_version ${hipcc_version})
   message(STATUS "HIP compiler version: ${hipcc_version}")
 
-  if(hipcc_version VERSION_LESS 6.4)
+  if(hipcc_version VERSION_LESS 6.3)
     message(FATAL_ERROR "You must use ROCm version >= 6.3 to compile Quokka. All previous ROCm versions have compiler bugs that may cause Quokka to produce incorrect results.")
   endif()
 endif(AMReX_GPU_BACKEND MATCHES "HIP")

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -64,7 +64,7 @@ if(AMReX_GPU_BACKEND MATCHES "HIP")
   endif()
 
   if(hipcc_version VERSION_LESS 6.3)
-    message(FATAL_ERROR "You must use ROCm version >= 6.3 to compile Quokka. All previous ROCm versions have compiler bugs that cause Quokka to produce incorrect results.")
+    message(FATAL_ERROR "You must use ROCm version >= 6.3 to compile Quokka. All previous ROCm versions have compiler bugs that may cause Quokka to produce incorrect results.")
   endif()
 endif(AMReX_GPU_BACKEND MATCHES "HIP")
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -64,7 +64,7 @@ if(AMReX_GPU_BACKEND MATCHES "HIP")
   string(REPLACE "HIP version: " "" hipcc_version ${hipcc_version})
   message(STATUS "HIP compiler version: ${hipcc_version}")
 
-  if(hipcc_version VERSION_LESS 6.3)
+  if(hipcc_version VERSION_LESS 6.4)
     message(FATAL_ERROR "You must use ROCm version >= 6.3 to compile Quokka. All previous ROCm versions have compiler bugs that may cause Quokka to produce incorrect results.")
   endif()
 endif(AMReX_GPU_BACKEND MATCHES "HIP")

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -59,9 +59,10 @@ if(AMReX_GPU_BACKEND MATCHES "HIP")
   )
   if(NOT hipcc_result EQUAL 0)
       message(FATAL_ERROR "Failed to detect HIP compiler version: ${hipcc_error}")
-  else()
-      message(STATUS "HIP compiler version: ${hipcc_version}")
   endif()
+
+  string(REPLACE "HIP version: " "" hipcc_version ${hipcc_version})
+  message(STATUS "HIP compiler version: ${hipcc_version}")
 
   if(hipcc_version VERSION_LESS 6.3)
     message(FATAL_ERROR "You must use ROCm version >= 6.3 to compile Quokka. All previous ROCm versions have compiler bugs that may cause Quokka to produce incorrect results.")

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -49,6 +49,25 @@ if(AMReX_GPU_BACKEND MATCHES "CUDA")
   endif(CMAKE_VERSION VERSION_LESS 3.20)
 endif(AMReX_GPU_BACKEND MATCHES "CUDA")
 
+if(AMReX_GPU_BACKEND MATCHES "HIP")
+  execute_process(
+      COMMAND hipcc --version
+      RESULT_VARIABLE hipcc_result
+      OUTPUT_VARIABLE hipcc_version
+      ERROR_VARIABLE hipcc_error
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+  if(NOT hipcc_result EQUAL 0)
+      message(FATAL_ERROR "Failed to detect HIP compiler version: ${hipcc_error}")
+  else()
+      message(STATUS "HIP compiler version: ${hipcc_version}")
+  endif()
+
+  if(hipcc_version VERSION_LESS 6.3)
+    message(FATAL_ERROR "You must use ROCm version >= 6.3 to compile Quokka. All previous ROCm versions have compiler bugs that cause Quokka to produce incorrect results.")
+  endif()
+endif(AMReX_GPU_BACKEND MATCHES "HIP")
+
 if(CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
   # abort! Intel Compiler Classic cannot compile Quokka!
   message(FATAL_ERROR "You have configured CMake to use the 'classic' Intel compilers (icc/icpc), which are the old Intel compilers. They cannot compile Quokka correctly! You must use the new LLVM-based Intel compilers (icx/icpx) instead by adding the following CMake command-line options: -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icpx")

